vscode/src/completions/providers/hot-streak.ts

```diff
@@ -0,0 +1,105 @@
+        updatedDocContext = insertIntoDocContext(updatedDocContext, lastInsertedWhitespace, languageId)
+
+        alreadyInsertedLength += completion.insertText.length + lastInsertedWhitespace.length
+
+        return updatedDocContext
+    }
+
+    function extract(rawCompletion: string, isRequestEnd: boolean): void {
+        do {
+            const unprocessedCompletion = rawCompletion.slice(alreadyInsertedLength)
+            if (unprocessedCompletion.length <= 0) {
+                break
+            }
+
+            const updatedProviderOptions = {
+                ...providerOptions,
+                docContext: updatedDocContext,
+            }
+
+            const eventualDynamicMultilineProviderOptions = providerOptions.dynamicMultilineCompletions
+                ? {
+                      ...updatedProviderOptions,
+                      docContext: getUpdatedDocContext({
+                          ...params,
+                          initialCompletion: unprocessedCompletion,
+                          completionPostProcessId: '',
+                      }),
+                  }
+                : updatedProviderOptions
+
+            const completion = canUsePartialCompletion(unprocessedCompletion, eventualDynamicMultilineProviderOptions)
+            if (completion) {
+                // If the partial completion logic finds a match, extract this as the next hot streak
+                const processedCompletion = processCompletion(completion, eventualDynamicMultilineProviderOptions)
+
+                onHotStreakCompletionReady(updatedDocContext, {
+                    ...processedCompletion,
+                    stopReason: 'hot-streak',
+                })
+
+                updatedDocContext = insertCompletionAndPressEnter(updatedDocContext, processedCompletion)
+            } else if (isRequestEnd) {
+                // If not and we are at processing the last payload, we use the whole remainder for
+                // the completion (this means we will parse the last line even when a \n is missing
+                // at the end)
+                const completion = parseAndTruncateCompletion(
+                    unprocessedCompletion,
+                    eventualDynamicMultilineProviderOptions
+                )
+                if (completion.insertText.trim().length <= 0) {
+                    break
+                }
+                const processedCompletion = processCompletion(completion, eventualDynamicMultilineProviderOptions)
+                onHotStreakCompletionReady(updatedDocContext, {
+                    ...processedCompletion,
+                    stopReason: 'hot-streak-end',
+                })
+                updatedDocContext = insertCompletionAndPressEnter(updatedDocContext, processedCompletion)
+            } else {
+                // If we don't have enough in the remaining completion text to generate a full
+                // hot-streak completion so we yield.
```

Comment:
We can have if we ... we yield, or we ... so we yield but if ... so doesn't go; so joins complete sentences.

```suggestion
                // We don't have enough in the remaining completion text to generate a full
                // hot-streak completion so we yield.
```